### DISCUSSION TITLE
How to add slice of widget to a Frame ?

### DISCUSSION THREAD

**Discussion by heejit** [cogentcore/core#1244] [Q&A] [answered]
In previous discussion we have created a card widget now I want to add slice of card widgets to a Frame ?

And on every new search slice will get updated also.

**Comment by heejit** [comment]
is this correct way :

```
vFrame11.DeleteChildren()
vFrame11.AddChild(this.CardWidget)
```

**Reply by kkoreilly** [reply]
You have two options: you can either use a `core.List` with your card registered as the `core.Value` type for the elements, or more simply you can do something like this, assuming you have some data slice:

```go
	data := []Data{}
	b.Maker(func(p *tree.Plan) {
		for i, d := range data {
			tree.AddAt(p, strconv.Itoa(i), func(w *Card) {
				w.SetHeading(d.Heading).SetSubHeading(d.SubHeading)
			})
		}
	})
```

This allows you to have it automatically update without manual deletion when you make new changes and call `Update`.

**Reply by heejit** [reply]
it works thanks.

**Reply by heejit** [reply]
If I don't call CardFrame.Update()/CardFrame.DeleteChildren() after clear the noteDataAry it shows the old data
is that okay because I am still calling Update after adding new data into noteDataAry

```
       // remove old
	this.noteDataAry = nil
	this.CardFrame.Update()

	// add new if nay
	for vRows.Next() {
		vRows.Scan(&vData.Id, &vData.Name, &vData.Time)
		this.noteDataAry = append(this.noteDataAry, vData)
	}
	this.CardFrame.Update()
```

This is maker call
```
	this.CardFrame.Maker(func(p *tree.Plan) {
		for i, vData := range this.noteDataAry {
			tree.AddAt(p, strconv.Itoa(i), func(w *Card) {
				w.SetHeading(vData.Name)
				w.SetSubHeading(fmt.Sprintf("ID=%d | LU-Time=%s", vData.Id, vData.Time))
				w.SetData(vData.Id)
				w.OnClick(func(e events.Event) {
					this.ShowNote(vData.Id)
				} )
			})
		}
	})
```

**Reply by kkoreilly** [reply, marked as answer]
The issue is that you need to re-fetch `vData` in each inner closure function inside of `AddAt`, since it is still the same function from before with a stale closure variable as long as the index is the same. So if you do something like this in your `Maker`, you shouldn't need the manual `Update` or `DeleteChildren` between clearing and adding:

```go
	for i := range this.noteDataAry {
		tree.AddAt(p, strconv.Itoa(i), func(w *Card) {
			w.Updater(func() {
				vData := this.noteDataAry[i]
				w.SetHeading(vData.Name)
				w.SetSubHeading(fmt.Sprintf("ID=%d | LU-Time=%s", vData.Id, vData.Time))
				w.SetData(vData.Id)
			})
			w.OnClick(func(e events.Event) {
				vData := this.noteDataAry[i]
				this.ShowNote(vData.Id)
			})
		})
	}
```

**Reply by heejit** [reply]
that works.
